Documentation about User Access Tokens (#477)
huggingface/huggingface_hub: docs/hub/security.md

# Security and trust on the Hugging Face Hub

## User Access Tokens

### What are User Access Tokens?

> User Access Tokens are the preferred way to authenticate an application or a notebook to the HuggingFace services. You can manage your access tokens in your [settings](https://huggingface.co/settings/tokens).

Reviewer: This link is a 404

```suggestion
User Access Tokens are the preferred way to authenticate an application or a notebook to the HuggingFace services. You can manage your access tokens in your [settings](https://huggingface.co/settings/token).
```

> User Access Tokens are the preferred way to authenticate an application or a notebook to the HuggingFace services. You can manage your access tokens in your [settings](https://huggingface.co/settings/tokens).

Reviewer 2: ```suggestion
User Access Tokens are the preferred way to authenticate an application or notebook to Hugging Face services. You can manage your access tokens in your [settings](https://huggingface.co/settings/tokens).
```

![/docs/assets/hub/access-tokens.png](/docs/assets/hub/access-tokens.png)

> Access tokens allow applications and notebooks to impersonate you, with restricted permissions. Each token is given a role, among the following:

Reviewer: ```suggestion
Access tokens allow applications and notebooks to impersonate you with restricted permissions. Each token is given a role, among the following:
```

> Access tokens allow applications and notebooks to impersonate you, with restricted permissions. Each token is given a role, among the following:

Reviewer 2: ```suggestion
Access tokens allow applications and notebooks to perform specific actions specified by the scope of the roles shown in the following:
```

> - `read`: tokens with this role can only be used to provide read access to repositories you could read. That includes public repositories and private repositories that you or an org you're a member of own. Use this role if you only need to read content from the HuggingFace hub (eg, when doing inference).

Reviewer: ```suggestion
- `read`: tokens with this role can only be used to provide read access to repositories you could read. That includes public repositories and private repositories that you or an org you're a member of own. Use this role if you only need to read content from the Hugging Face Hub (e.g. when downloading private models or doing inference).
```

> - `read`: tokens with this role can only be used to provide read access to repositories you could read. That includes public repositories and private repositories that you or an org you're a member of own. Use this role if you only need to read content from the HuggingFace hub (eg, when doing inference).

Reviewer 2: ```suggestion
- `read`: tokens with this role can only be used to provide read access to repositories you could read. That includes public repositories and private repositories that you, or an organization you're a member of, own. Use this role if you only need to read content from the HuggingFace hub (eg, when doing inference).
```

> - `read`: tokens with this role can only be used to provide read access to repositories you could read. That includes public repositories and private repositories that you or an org you're a member of own. Use this role if you only need to read content from the Hugging Face Hub (e.g. when downloading private models or doing inference).

Reviewer 3: ```suggestion
- `read`: tokens with this role can only be used to provide read access to repositories you could read. That includes public repositories and private repositories that you or one of your orgs own. Use this role if you only need to read content from the Hugging Face Hub (e.g. when downloading private models or doing inference).
```
maybe simpler grammatically? not sure

> - `write`: tokens with this role additionally grant write access to the repositories you have write access to. Use this token if you need to push content to a repository (eg, when training a model).

Reviewer 4: ```suggestion
- `write`: tokens with this role additionally grant write access to the repositories you have write access to. Use this token if you need to create or push content to a repository (eg, when training a model).
```

> 

Reviewer 4: No doc for the admin token?

Author: We are discussing wether or not we want to keep it, as it's essentially useless now - but could grant programmatic access to organizations' settings in the future

> - `write`: tokens with this role additionally grant write access to the repositories you have write access to. Use this token if you need to create or push content to a repository (eg, when training a model).

Reviewer: ```suggestion
- `write`: tokens with this role additionally grant write access to the repositories you have write access to. Use this token if you need to create or push content to a repository (e.g., when training a model or modifying a model card).
```

### How to manage User Access Tokens?

To create an access token, go to your settings, then click on the ["Access Tokens" tab](https://huggingface.co/settings/token). Click on the "New token" button to create a new User Access Token.

![/docs/assets/hub/new-token.png](/docs/assets/hub/new-token.png)

Select a role and a name for your token and voilà - you're ready to go!

> You can delete an refresh User Access Tokens by clicking on the "Manage" button.

Reviewer 2: ```suggestion
You can delete and refresh User Access Tokens by clicking on the "Manage" button.
```

![/docs/assets/hub/delete-token.png](/docs/assets/hub/delete-token.png)

### How to use User Access Tokens?

> There are plenty of ways to use a User Access Token to access the Hugging Face hub, granting you the flexibility you need to build awesome apps on top of it.

Reviewer: ```suggestion
There are plenty of ways to use a User Access Token to access the Hugging Face Hub, granting you the flexibility you need to build awesome apps on top of it.
```

> - User Access Tokens can be used **in place of a password** to access the Hugging Face hub with git or with Basic authentication

Reviewer: ```suggestion
- User Access Tokens can be used **in place of a password** to access the Hugging Face Hub with git or with basic authentication
```

> - User Access Tokens can be passed as a **bearer token** when calling the Hugging Face APIs

Reviewer: ```suggestion
- User Access Tokens can be passed as a **bearer token** when calling the [Inference API](https://huggingface.co/inference-api).
```

> - User Access Tokens can be used in the Hugging Face Python library:

Reviewer: Not 100% sure we should link to that page

```suggestion
- User Access Tokens can be used in the `huggingface_hub` [Python library](https://github.com/huggingface/huggingface_hub/tree/main/src/huggingface_hub):
```

Author: Shoot I wanted to say "libraries", to refer to `datasets`, `transformers` and `huggingface_hub` all at once

```python
from transformers import AutoModel

access_token = "hf_..."

model = AutoModel.from_pretrained("private/model", use_auth_token=access_token)
```

> 

Reviewer 3: ```suggestion

⚠️ Try not to leak your token! Though you can always rotate it, anyone will be able to read or write your private repos in the meantime which is 💩 
```

### Best practices

> We recommand to create one access token per app or usage. For instance, one would have one token to use on their local machine, one token to use on Colab notebooks, and one token to be used by an awesome custom inference server. This way, you can invalidate one token without impacting your other usages.

Reviewer 2: ```suggestion
We recommend you create one access token per app or usage. For instance, you could have a separate token for:

 * A local machine.
 * A Colab notebook.
 * An awesome custom inference server. 
 
 This way, you can invalidate one token without impacting your other usages.
```

> We also recommand giving the minimal rights to each token you create. If you only need read access (ie, loading a dataset with the `datasets` library or retrieving the weights of a model), give your access token the `read` role.

Reviewer 4: ```suggestion
We also recommand giving the minimal rights to each token you create. If you only need read access (i.e., loading a dataset with the `datasets` library or retrieving the weights of a model), give your access token the `read` role.
```

> We also recommand giving the minimal rights to each token you create. If you only need read access (i.e., loading a dataset with the `datasets` library or retrieving the weights of a model), give your access token the `read` role.

Reviewer 2: ```suggestion
We also recommend only giving the appropriate role to each token you create. If you only need read access (i.e., loading a dataset with the `datasets` library or retrieving the weights of a model), only give your access token the `read` role.
```
